src/Engine/EmitMSIL/CodeGenIL.cs

```diff
@@ -1405,26 +1407,111 @@ private Type EmitFunctionCallNode(AssociativeNode node)
+
+                EmitOpCode(OpCodes.Pop);
+                if (local != null)
+                {
+                    EmitOpCode(OpCodes.Ldloc, local.LocalIndex);
+                }
+
+                //emit an array of args to pass to unmarshal
+                EmitArray(typeof(object), args, (AssociativeNode n, int index) =>
+                {
+                    Type t = DfsTraverse(n);
                     if (isStaticOrCtor)
                     {
-                        t = EmitCoercionCode(arg, t, parameters[index]);
+                        t = EmitCoercionCode(n, t, parameters[index]);
                     }
                     else if (index > 0)
                     {
-                        t = EmitCoercionCode(arg, t, parameters[index - 1]);
+                        t = EmitCoercionCode(n, t, parameters[index - 1]);
+                    }
+
+                    if (t == null) return;
+
+                    if (t.IsValueType)
+                    {
+                        EmitOpCode(OpCodes.Box, t);
+                    }
+                });
+
+                //TODO if possible only marshal args that resulted from a replicated call.
+                
+                //emit marshal for args
+
+                // Emit call to load the runtimeCore argument
+                EmitOpCode(OpCodes.Ldarg_3);
+                //TODO cache this at start.
+                //now call marshall to convert any clrstackvalues to plain c# objs. 
+                var marhshallMethod = typeof(Replication).GetMethod(nameof(Replication.UnMarshalFunctionArguments2), BindingFlags.Static | BindingFlags.Public);
+                EmitOpCode(OpCodes.Call, marhshallMethod);
+
+                var unMarshaledArgsArray = DeclareLocal(marhshallMethod.ReturnType, "unMarshaledArgsArray");
+                if (unMarshaledArgsArray != null)
+                {
+                    EmitOpCode(OpCodes.Stloc, unMarshaledArgsArray.LocalIndex);
+                }
+                //now on the stack we have a list of objects, we need to iterate through it and push each item to the stack.
+                //emit a for loop which pushes to stack.
+
+                //foreach item in args -
+                //increment index
+                //emit code to index into array
+                //ldelem ref.
+                // we know array.length because it must be the same length as args.len
+
+                if (unMarshaledArgsArray != null)
+                {
+                    for (int i = 0; i < args.Count; i++)
+                    {
+                        //load array
+                        EmitOpCode(OpCodes.Ldloc, unMarshaledArgsArray.LocalIndex);
+                        EmitOpCode(OpCodes.Ldc_I4, i);
+                        EmitOpCode(OpCodes.Ldelem_Ref);
+                        if (parameters[i].IsValueType)
+                        {
+                            EmitOpCode(OpCodes.Unbox_Any, parameters[i]);
+                        }
```

Comment:
@mjkkirschner, if I understand this correctly, shouldn't this be something like:
```
EmitOpCode(OpCodes.Ldloc, unMarshaledArgsArray.LocalIndex);
 EmitOpCode(OpCodes.Ldc_I4, i);
if (parameters[i].IsValueType)
{
            EmitOpCode(OpCodes.Unbox_Any, parameters[i]);
            EmitOpCode(OpCodes.Ldelem, typeof(parameters[i]));
}
else EmitOpCode(OpCodes.Ldelem_Ref);

Reply:
in your sample code, the object reference is not yet on the stack, wouldn't that unbox be applied to the constant `i` - not the value in the array?

Reply:
I'm not sure how unbox works but how about this:
```
EmitOpCode(OpCodes.Ldloc, unMarshaledArgsArray.LocalIndex);
 EmitOpCode(OpCodes.Ldc_I4, i);
EmitOpCode(OpCodes.Ldelem_Ref);
if (parameters[i].IsValueType)
{
            EmitOpCode(OpCodes.Unbox_Any, parameters[i]);
            EmitOpCode(OpCodes.Ldelem, typeof(parameters[i]));
}
```

Reply:
Also, what's the difference between unbox_any and unbox?

Reply:
Ok, I was wondering if a `ldelem` was required after the unboxing, so this clears it.